django/contrib/admin/templatetags/admin_list.py

```diff
@@ -340,19 +340,33 @@ def date_hierarchy(cl):
         year_lookup = cl.params.get(year_field)
         month_lookup = cl.params.get(month_field)
         day_lookup = cl.params.get(day_field)
+        approximate = cl.approximate_date_hierarchy
 
         def link(filters):
             return cl.get_query_string(filters, [field_generic])
 
+        def minmax(qs):
+            date_range = qs.aggregate(first=models.Min(field_name),
+                                      last=models.Max(field_name))
+            return date_range['first'], date_range['last']
+
+
         if not (year_lookup or month_lookup or day_lookup):
             # select appropriate start level
             date_range = cl.queryset.aggregate(first=models.Min(field_name),
                                                last=models.Max(field_name))
-            if date_range['first'] and date_range['last']:
-                if date_range['first'].year == date_range['last'].year:
-                    year_lookup = date_range['first'].year
-                    if date_range['first'].month == date_range['last'].month:
-                        month_lookup = date_range['first'].month
+            # if date_range['first'] and date_range['last']:
```

Comment:
I'll add docs
I'll try to add tests. at version 1.11 they were not work .
at version 1.9 I don't try to run it